pennylane/transforms/core/transform_dispatcher.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the transform function, the transform dispatcher and the transform container.
+"""
+import copy
+import pennylane as qml
+
+
+class TransformError(Exception):
+    """Raised when there is an error with the transform logic"""
+
+
+class TransformDispatcher:
+    r"""This object is developer facing and should not be used directly to create transforms. Use
+    :func:`~.pennylane.transforms.core.transform`.
+
+    Convert a transform that has the signature (tape -> Sequence(tape), fn) to a transform dispatcher that can act
+    on tape, qfunc and qnode.
+
+    .. warning::
+
+        This class is developer-facing and should not be used directly.
+
+    .. seealso:: :func:`~.pennylane.transforms.core.transform`
+    """
+
+    def __init__(
+        self, transform, expand_transform=None, classical_cotransform=None
+    ):  # pylint:disable=redefined-outer-name
+        self._transform = transform
+        self._expand_transform = expand_transform
+        self._classical_cotransform = classical_cotransform
+
+    def __call__(self, *targs, **tkwargs):
+        obj = None
+
+        if targs:
+            # assume the first argument passed to the transform
+            # is the object we wish to transform
+            obj, *targs = targs
+
+        if isinstance(obj, qml.tape.QuantumTape):
+            new_tape = copy.deepcopy(obj)
+            return self._transform(new_tape, *targs, **tkwargs)
```

Comment:
I wonder if there's a way we can add a constraint forcing transforms to be purely functional. Right now the `expand` function modifies its input, but that's a problem we need to solve with `expand`, not something we want to actually allow functions to do.